Add E2E testing guidance using Postman and Newman (#900)
microsoft/code-with-engineering-playbook: docs/automated-testing/e2e-testing/recipes/postman-testing.md

> # Postman and Newman

Reviewer: can you add an overview that describes the flow so that the reader understands what this is about and can decide whether to continue reading or not

Author: Pushed.

## What is an Environment File

A Postman Environment file holds environment variables that can be referenced by a valid Postman Collection.

## What is a Postman Script

A Postman Script is Javascript hosted within a Postman Collection that can be written to execute against your Postman Collection and Environment File.

## What is the Newman Run Command

A Newman CLI command that allows you to specify a Postman Collection to be run.

## Installing Postman and Newman

For specific instruction on installing Postman, visit the [Downloads Postman](https://www.postman.com/downloads/) page.

For specific instruction on installing Newman, visit the [NPMJS Newman package](https://www.npmjs.com/package/newman) page.

## Implementing Automated End-to-end (E2E) Tests With Postman Collections

In order to provide guidance on implementing E2E automated tests with Postman, the section below begins with a use case that explains the trade-offs a dev or qa analyst might face when intending to use Postman for early testing. Each use case represents scenarios that facilitate the end goal of automated E2E tests.

### Use Case - Hands-on Functional Testing Of Endpoints

> A developer or qa analyst would like to locally test input data against API services that all share a common oauth2 token. As a result, they use Postman to craft an API test suite of Postman Collections that can be locally executed against individual endpoints across environments. After validating that their Postman Collection works, they share it with their team.

Reviewer 2: ```suggestion
A developer or QA analyst would like to locally test input data against API services that all share a common oauth2 token. As a result, they use Postman to craft an API test suite of Postman Collections that can be locally executed against individual endpoints across environments. After validating that their Postman Collection works, they share it with their team.
```

- You've set yourself up for the beginning stages of an E2E postman collection by aggregating the collections into a single file and using environment files to make it easier to switch environments.
- Token is refreshed automatically on every call in the collection. This saves you time normally lost from manually having to request a token that expired.
- Grants QA/Dev granular control of submitting combinations of input data per endpoint.
- Grants devs a common experience via Postman IDE features.

Ending with this approach has the following downsides:

- Promotes unsafe sharing of secrets. Credentials needed to request JWT token in the prefetch script are being manually shared.
- Secrets may happen to get exposed in the git commit history for various reasons (ex. Sharing the exported Postman Environment files).
- Collections can only be used locally to hit APIs (local or deployed). Not CI based.
- Each developer has to keep both their Postman Collection and Postman environment file(s) updated in order to keep up with latest changes to deployed services.

### Use Case - Hands-on Functional Testing Of Endpoints with Azure Key Vault and Azure App Config

> A developer or qa analyst may have an existing API test suite of Postman Collections, however, they now want to discourage unsafe sharing of secrets. As a result, they build a script that connects to both Keyvault and Azure App Config in order to automatically generate Postman Environment files instead of checking them into a shared repository.

Reviewer 2: ```suggestion
A developer or QA analyst may have an existing API test suite of Postman Collections, however, they now want to discourage unsafe sharing of secrets. As a result, they build a script that connects to both Key Vault and Azure App Config in order to automatically generate Postman Environment files instead of checking them into a shared repository.
```

    envFile = @{ "_postman_variable_scope" = "environment", "name" = $env, values = @() }
    foreach($var in $envVars){
            $envFile.values += @{ key = $var.key; value = $var.value; }
    }
    $envFile | ConvertTo-Json -depth 50 | Out-File -encoding ASCII -FilePath .\$env.postman_environment.json
    ```

6. Use Postman IDE to import the Postman Environment files to be referenced by your collection.

This approach has the following upsides:

- Inherits all the upsides of the previous case.
- Discourages unsafe sharing of secrets. Secrets are now pulled from Key Vault via Azure CLI. Keyvault Uri also no longer needs to be shared for access to auth tokens.
- Single source of truth for Postman Environment files. There's no longer a need to share them via repo.
- Developer only has to manage a single Postman Collection.

Ending with this approach has the following downsides:

- Secrets may happen to get exposed in the git commit history if .gitIgnore is not updated to ignore Postman Environment files.
- Collections can only be used locally to hit APIs (local or deployed). Not CI based.

### Use Case - E2E testing With Continuous Integration and Newman

> A developer or qa analyst may have an existing API test suite of local Postman Collections that follow security best practices for development, however, they now want E2E tests to run as part of automated CI pipeline. With the advent of Newman, you can now more readily use Postman to craft an API test suite executable in your CI.

Reviewer 2: ```suggestion
A developer or QA analyst may have an existing API test suite of local Postman Collections that follow security best practices for development, however, they now want E2E tests to run as part of automated CI pipeline. With the advent of Newman, you can now more readily use Postman to craft an API test suite executable in your CI.
```

    foreach($var in $envVars){
            $envFile.values += @{ key = $var.key; value = $var.value; }
    }
    $envFile | ConvertTo-Json -depth 50 | Out-File -encoding ASCII $env.postman_environment.json
    # 8. install Newman
    npm install --save-dev newman
    # 9. run automated E2E tests via Newman
    node_modules\.bin\newman run tests\e2e_Postman_collection.json -e $env.postman_environment.json
    ```

4. Create a yaml file and define a step that will run your test script. (ex. A yaml file targeting Azure Devops that runs a Powershell script.)

    ```yaml
    # Please treat this as pseudocode, and adjust as necessary.
    ############################################################
    displayName: 'Run Postman E2E tests'
    inputs:
        targetType: 'filePath'
        filePath: RunPostmanE2eTests.ps1
    env:
        APP_ID: $(environment.appId) # credentials for az cli
        AZURE_SECRET: $(environment.secret)
        AZURE_TENANT: $(environment.tenant)
    ```

This approach has the following upside:

- E2E tests can now be run automatically as part of a CI pipeline.

Ending with this approach has the following downside:

> - Postman Environment files are no longer being output to a local environment for hands-on manual testing. However, this can be solved by managing 2 scripts. 1 script to output environment variables locally and 1 for ci. Of course, reintroducing the script that outputs environment files invites the opportunity for files with secrets to get checked into git.

Reviewer 2: ```suggestion
- Postman Environment files are no longer being output to a local environment for hands-on manual testing. However, this can be solved by managing 2 scripts. 1 script to output environment variables locally and 1 for CI. Of course, reintroducing the script that outputs environment files invites the opportunity for files with secrets to get checked into git.
```
